[Logo](https://www.kitepackaging.co.uk/)
<p style="background-color:#808080;"><span style="color:#ffffff;"><i> 
<img src="https://www.kitepackaging.co.uk/images/kite-logo-2021.png" alt="Logo" width="100" height="200">

>>>>#                            **KITE PACKAGING : CLEANING OF DATA SETS**

#

<div style="text-align: center;">
    <img src="https://media.kitepackaging.co.uk/images/content/kite-fleet-6df6ac59-afec-490a-b826-5ccf19a1d715..png?width=2400&optimize=medium" alt="Logo" width="800" height="700">
</div>
'''


[Logo](https://www.kitepackaging.co.uk/)
<p style="background-color:#808080;"><span style="color:#ffffff;"><i> 
<img src="https://www.kitepackaging.co.uk/images/kite-logo-2021.png" alt="Logo" width="100" height="200">

#

In [31]:
# Import libraries and packages.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Ignore warnings.
import warnings
warnings.filterwarnings('ignore')

>**Note**: 'postcodes_coord_distance.csv' file was created using coordinates and distances generated with the Postcodes.io API and Google Matrix API. For detailed steps, please refer to the 'Generating Distances from Partial Postcodes.ipynb' notebook.

In [33]:
# Load and store data in DataFrames.
order_details = pd.read_excel('order_details.xlsx')
orders = pd.read_excel('orders.xlsx')
customers = pd.read_csv('postcodes_coord_distance.csv') 

In [35]:
# Make a copy of the raw data.
order_details_raw = order_details.copy()
orders_raw =  orders.copy()
customers_raw = customers.copy()

#

[Logo](https://www.kitepackaging.co.uk/)
<p style="background-color:#808080;"><span style="color:#ffffff;"><i> 
<img src="https://www.kitepackaging.co.uk/images/kite-logo-2021.png" alt="Logo" width="100" height="200">

### **1 - Order details**

In [12]:
# View the first 5 rows of the dataset.
order_details.head()

OrderID           PartNumber  \
0  2729772            BB305X435   
1  2729773              PBL1215   
2  2729774   CSW222X150X88-BPB6   
3  2729774  PS320X140X420-BROWN   
4  2729774           TPK048050S   

                                         Description  Packs  PackPrice  \
0          Clear bubble bags 305x435mm - pack of 150      1      26.46   
1  Light duty polythene bags - 11.75x14.75 inches...      1      18.80   
2  Brown postal boxes - 222x150x88mm (8.75x6x3.5 ...      1      20.55   
3  Twist handle paper carrier bags brown 320x140x...      1      52.28   
4  Paper kraft tape - 48mm x 50m, Solvent - pack ...      3      13.05   

   UnitsPerPack  PackWeight  LineWeight  PalletQuantity  CubicMPerUnit  
0           150        2.55        2.55            2400       0.000925  
1          1000        4.40        4.40          180000       0.000006  
2            50        3.90        3.90            2200       0.000726  
3           250        5.00        5.00            3750       0.000144  
4             6        1.28        3.84            2592       0.000677

In [13]:
# Display summary of DataFrame.
order_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76573 entries, 0 to 76572
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   OrderID         76573 non-null  int64  
 1   PartNumber      76573 non-null  object 
 2   Description     76573 non-null  object 
 3   Packs           76573 non-null  int64  
 4   PackPrice       76573 non-null  float64
 5   UnitsPerPack    76573 non-null  int64  
 6   PackWeight      76573 non-null  float64
 7   LineWeight      76402 non-null  float64
 8   PalletQuantity  76573 non-null  int64  
 9   CubicMPerUnit   72449 non-null  float64
dtypes: float64(4), int64(4), object(2)
memory usage: 5.8+ MB


In [14]:
# Count the total number of missing values in the DataFrame
null_values_count = order_details.isnull().sum().sum()

# Get columns with missing values
columns_with_null = order_details.columns[order_details.isnull().any()].tolist()

# Format the output message
output_message = f"The total number of null values in the DataFrame is: {null_values_count}\n"
if columns_with_null:
    output_message += f"Missing values found in the following columns: {', '.join(columns_with_null)}"
else:
    output_message += "No columns have null values."

print(output_message)

The total number of null values in the DataFrame is: 4295
Missing values found in the following columns: LineWeight, CubicMPerUnit


In [15]:
# Look for null values.
order_details.isnull().sum()

OrderID              0
PartNumber           0
Description          0
Packs                0
PackPrice            0
UnitsPerPack         0
PackWeight           0
LineWeight         171
PalletQuantity       0
CubicMPerUnit     4124
dtype: int64

> There are **171 missing values** in column 'LineWeight'
>
> There are **4124 missing values** in column 'CubicMperUnit'
>
> `Please note that:`**Since the removal of these values could impact the analysis of some of the metrics used, it has been agreed that none of the values will be removed.**

In [17]:
# Rename columns in a more simple manner.
order_details = order_details.rename(columns={
    'OrderID': 'order_id',
    'PartNumber': 'part_number',
    'Description': 'description',
    'Packs': 'packs',
    'PackPrice': 'pack_price',
    'UnitsPerPack': 'units_per_pack',
    'PackWeight': 'pack_weight',
    'LineWeight': 'line_weight',
    'PalletQuantity': 'pallet_quantity',
    'CubicMPerUnit': 'cubic_m_per_unit'
})

# See updated DataFrame.
order_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76573 entries, 0 to 76572
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          76573 non-null  int64  
 1   part_number       76573 non-null  object 
 2   description       76573 non-null  object 
 3   packs             76573 non-null  int64  
 4   pack_price        76573 non-null  float64
 5   units_per_pack    76573 non-null  int64  
 6   pack_weight       76573 non-null  float64
 7   line_weight       76402 non-null  float64
 8   pallet_quantity   76573 non-null  int64  
 9   cubic_m_per_unit  72449 non-null  float64
dtypes: float64(4), int64(4), object(2)
memory usage: 5.8+ MB


>### **Add additional columns**

>>**total_pack_price:**

In [20]:
# Calculate 'total_pack_price'.
order_details ['total_pack_price'] = order_details ['packs']* order_details['pack_price']
# See updated DataFrame.
order_details.head()

order_id          part_number  \
0   2729772            BB305X435   
1   2729773              PBL1215   
2   2729774   CSW222X150X88-BPB6   
3   2729774  PS320X140X420-BROWN   
4   2729774           TPK048050S   

                                         description  packs  pack_price  \
0          Clear bubble bags 305x435mm - pack of 150      1       26.46   
1  Light duty polythene bags - 11.75x14.75 inches...      1       18.80   
2  Brown postal boxes - 222x150x88mm (8.75x6x3.5 ...      1       20.55   
3  Twist handle paper carrier bags brown 320x140x...      1       52.28   
4  Paper kraft tape - 48mm x 50m, Solvent - pack ...      3       13.05   

   units_per_pack  pack_weight  line_weight  pallet_quantity  \
0             150         2.55         2.55             2400   
1            1000         4.40         4.40           180000   
2              50         3.90         3.90             2200   
3             250         5.00         5.00             3750   
4               6         1.28         3.84             2592   

   cubic_m_per_unit  total_pack_price  
0          0.000925             26.46  
1          0.000006             18.80  
2          0.000726             20.55  
3          0.000144             52.28  
4          0.000677             39.15

>>**total_units:**

In [22]:
# Calculate 'total_units'
order_details ['total_units'] = order_details ['packs']* order_details['units_per_pack']
# See updated DataFrame.
order_details.head()

order_id          part_number  \
0   2729772            BB305X435   
1   2729773              PBL1215   
2   2729774   CSW222X150X88-BPB6   
3   2729774  PS320X140X420-BROWN   
4   2729774           TPK048050S   

                                         description  packs  pack_price  \
0          Clear bubble bags 305x435mm - pack of 150      1       26.46   
1  Light duty polythene bags - 11.75x14.75 inches...      1       18.80   
2  Brown postal boxes - 222x150x88mm (8.75x6x3.5 ...      1       20.55   
3  Twist handle paper carrier bags brown 320x140x...      1       52.28   
4  Paper kraft tape - 48mm x 50m, Solvent - pack ...      3       13.05   

   units_per_pack  pack_weight  line_weight  pallet_quantity  \
0             150         2.55         2.55             2400   
1            1000         4.40         4.40           180000   
2              50         3.90         3.90             2200   
3             250         5.00         5.00             3750   
4               6         1.28         3.84             2592   

   cubic_m_per_unit  total_pack_price  total_units  
0          0.000925             26.46          150  
1          0.000006             18.80         1000  
2          0.000726             20.55           50  
3          0.000144             52.28          250  
4          0.000677             39.15           18

>>**pallet_space:**

In [24]:
# Calculate 'pallet_space'
# Use apply() to set 'pallet_space' based on a condition
order_details['pallet_space'] = order_details.apply(lambda row: 
                                                    row['total_units'] / row['pallet_quantity']\
                                                    if row['pallet_quantity'] != 0 else 0,axis=1)

# See updated DataFrame.
order_details.head()

order_id          part_number  \
0   2729772            BB305X435   
1   2729773              PBL1215   
2   2729774   CSW222X150X88-BPB6   
3   2729774  PS320X140X420-BROWN   
4   2729774           TPK048050S   

                                         description  packs  pack_price  \
0          Clear bubble bags 305x435mm - pack of 150      1       26.46   
1  Light duty polythene bags - 11.75x14.75 inches...      1       18.80   
2  Brown postal boxes - 222x150x88mm (8.75x6x3.5 ...      1       20.55   
3  Twist handle paper carrier bags brown 320x140x...      1       52.28   
4  Paper kraft tape - 48mm x 50m, Solvent - pack ...      3       13.05   

   units_per_pack  pack_weight  line_weight  pallet_quantity  \
0             150         2.55         2.55             2400   
1            1000         4.40         4.40           180000   
2              50         3.90         3.90             2200   
3             250         5.00         5.00             3750   
4               6         1.28         3.84             2592   

   cubic_m_per_unit  total_pack_price  total_units  pallet_space  
0          0.000925             26.46          150      0.062500  
1          0.000006             18.80         1000      0.005556  
2          0.000726             20.55           50      0.022727  
3          0.000144             52.28          250      0.066667  
4          0.000677             39.15           18      0.006944

In [25]:
# Display updated DataFrame.
order_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76573 entries, 0 to 76572
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          76573 non-null  int64  
 1   part_number       76573 non-null  object 
 2   description       76573 non-null  object 
 3   packs             76573 non-null  int64  
 4   pack_price        76573 non-null  float64
 5   units_per_pack    76573 non-null  int64  
 6   pack_weight       76573 non-null  float64
 7   line_weight       76402 non-null  float64
 8   pallet_quantity   76573 non-null  int64  
 9   cubic_m_per_unit  72449 non-null  float64
 10  total_pack_price  76573 non-null  float64
 11  total_units       76573 non-null  int64  
 12  pallet_space      76573 non-null  float64
dtypes: float64(6), int64(5), object(2)
memory usage: 7.6+ MB


>>**Recalculate column 'line_weight':**

In [27]:
# Calculate 'Line_weight'
order_details['line_weight_recalc'] = order_details ['packs']* order_details['pack_weight']

# See updated DataFrame.
order_details.head()

order_id          part_number  \
0   2729772            BB305X435   
1   2729773              PBL1215   
2   2729774   CSW222X150X88-BPB6   
3   2729774  PS320X140X420-BROWN   
4   2729774           TPK048050S   

                                         description  packs  pack_price  \
0          Clear bubble bags 305x435mm - pack of 150      1       26.46   
1  Light duty polythene bags - 11.75x14.75 inches...      1       18.80   
2  Brown postal boxes - 222x150x88mm (8.75x6x3.5 ...      1       20.55   
3  Twist handle paper carrier bags brown 320x140x...      1       52.28   
4  Paper kraft tape - 48mm x 50m, Solvent - pack ...      3       13.05   

   units_per_pack  pack_weight  line_weight  pallet_quantity  \
0             150         2.55         2.55             2400   
1            1000         4.40         4.40           180000   
2              50         3.90         3.90             2200   
3             250         5.00         5.00             3750   
4               6         1.28         3.84             2592   

   cubic_m_per_unit  total_pack_price  total_units  pallet_space  \
0          0.000925             26.46          150      0.062500   
1          0.000006             18.80         1000      0.005556   
2          0.000726             20.55           50      0.022727   
3          0.000144             52.28          250      0.066667   
4          0.000677             39.15           18      0.006944   

   line_weight_recalc  
0                2.55  
1                4.40  
2                3.90  
3                5.00  
4                3.84

In [28]:
# Display updated DataFrame.
order_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76573 entries, 0 to 76572
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            76573 non-null  int64  
 1   part_number         76573 non-null  object 
 2   description         76573 non-null  object 
 3   packs               76573 non-null  int64  
 4   pack_price          76573 non-null  float64
 5   units_per_pack      76573 non-null  int64  
 6   pack_weight         76573 non-null  float64
 7   line_weight         76402 non-null  float64
 8   pallet_quantity     76573 non-null  int64  
 9   cubic_m_per_unit    72449 non-null  float64
 10  total_pack_price    76573 non-null  float64
 11  total_units         76573 non-null  int64  
 12  pallet_space        76573 non-null  float64
 13  line_weight_recalc  76573 non-null  float64
dtypes: float64(7), int64(5), object(2)
memory usage: 8.2+ MB


In [29]:
# Delete unnessary columns
order_details = order_details.drop(['line_weight'], axis= 1)

# View Dataframe
order_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76573 entries, 0 to 76572
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            76573 non-null  int64  
 1   part_number         76573 non-null  object 
 2   description         76573 non-null  object 
 3   packs               76573 non-null  int64  
 4   pack_price          76573 non-null  float64
 5   units_per_pack      76573 non-null  int64  
 6   pack_weight         76573 non-null  float64
 7   pallet_quantity     76573 non-null  int64  
 8   cubic_m_per_unit    72449 non-null  float64
 9   total_pack_price    76573 non-null  float64
 10  total_units         76573 non-null  int64  
 11  pallet_space        76573 non-null  float64
 12  line_weight_recalc  76573 non-null  float64
dtypes: float64(6), int64(5), object(2)
memory usage: 7.6+ MB


>>**total_cubic_m:**

In [31]:
# Calculate 'total_cubic_m'.
order_details['total_cubic_m'] = order_details ['total_units']* order_details['cubic_m_per_unit']

# See updated DataFrame.
order_details.head()

order_id          part_number  \
0   2729772            BB305X435   
1   2729773              PBL1215   
2   2729774   CSW222X150X88-BPB6   
3   2729774  PS320X140X420-BROWN   
4   2729774           TPK048050S   

                                         description  packs  pack_price  \
0          Clear bubble bags 305x435mm - pack of 150      1       26.46   
1  Light duty polythene bags - 11.75x14.75 inches...      1       18.80   
2  Brown postal boxes - 222x150x88mm (8.75x6x3.5 ...      1       20.55   
3  Twist handle paper carrier bags brown 320x140x...      1       52.28   
4  Paper kraft tape - 48mm x 50m, Solvent - pack ...      3       13.05   

   units_per_pack  pack_weight  pallet_quantity  cubic_m_per_unit  \
0             150         2.55             2400          0.000925   
1            1000         4.40           180000          0.000006   
2              50         3.90             2200          0.000726   
3             250         5.00             3750          0.000144   
4               6         1.28             2592          0.000677   

   total_pack_price  total_units  pallet_space  line_weight_recalc  \
0             26.46          150      0.062500                2.55   
1             18.80         1000      0.005556                4.40   
2             20.55           50      0.022727                3.90   
3             52.28          250      0.066667                5.00   
4             39.15           18      0.006944                3.84   

   total_cubic_m  
0       0.138750  
1       0.006000  
2       0.036300  
3       0.036000  
4       0.012186

In [32]:
# View Dataframe
order_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76573 entries, 0 to 76572
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            76573 non-null  int64  
 1   part_number         76573 non-null  object 
 2   description         76573 non-null  object 
 3   packs               76573 non-null  int64  
 4   pack_price          76573 non-null  float64
 5   units_per_pack      76573 non-null  int64  
 6   pack_weight         76573 non-null  float64
 7   pallet_quantity     76573 non-null  int64  
 8   cubic_m_per_unit    72449 non-null  float64
 9   total_pack_price    76573 non-null  float64
 10  total_units         76573 non-null  int64  
 11  pallet_space        76573 non-null  float64
 12  line_weight_recalc  76573 non-null  float64
 13  total_cubic_m       72449 non-null  float64
dtypes: float64(7), int64(5), object(2)
memory usage: 8.2+ MB


In [33]:
# Look for null values.
order_details.isnull().sum()

order_id                 0
part_number              0
description              0
packs                    0
pack_price               0
units_per_pack           0
pack_weight              0
pallet_quantity          0
cubic_m_per_unit      4124
total_pack_price         0
total_units              0
pallet_space             0
line_weight_recalc       0
total_cubic_m         4124
dtype: int64

In [34]:
# Save DataFrame as csv.
order_details.to_csv('order_details_clean.csv')

#

[Logo](https://www.kitepackaging.co.uk/)
<p style="background-color:#808080;"><span style="color:#ffffff;"><i> 
<img src="https://www.kitepackaging.co.uk/images/kite-logo-2021.png" alt="Logo" width="100" height="200">

### **2 - Orders**

In [38]:
# View the first few rows of the dataset.
orders.head()

OrderID  WebsiteID               OrderDate  StatusName  StatusId  \
0  2729772     971172 2024-06-03 00:15:07.230  Dispatched         6   
1  2729773     971173 2024-06-03 00:33:45.437  Dispatched         6   
2  2729774     889210 2024-06-03 00:52:10.663  Dispatched         6   
3  2729775     971174 2024-06-03 00:54:54.100  Dispatched         6   
4  2729776     971175 2024-06-03 01:19:04.583  Dispatched         6   

  PaymentOptionName  PaymentType DespatchedCourier  ProductsExVAT  \
0            Paypal            1        DX Express          26.46   
1   Payment Gateway            2        DX Express          18.80   
2   Payment Gateway            2        DX Freight         351.07   
3   Payment Gateway            2        DX Express          56.14   
4            Paypal            1        DX Freight         144.16   

   DeliveryExVAT    VAT  SubTotal  GrandTotal DateDespatched  DeliveryCost  
0           7.37   6.76     26.46       40.59     2024-06-03          6.95  
1           7.37   5.23     18.80       31.40     2024-06-03          6.95  
2          19.99  74.22    351.07      445.28     2024-06-03         25.33  
3           0.00  11.23     56.14       67.37     2024-06-03          6.95  
4           0.00  28.84    144.16      173.00     2024-06-03         16.87

In [39]:
# Display summary of DataFrame.
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32063 entries, 0 to 32062
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   OrderID            32063 non-null  int64         
 1   WebsiteID          32063 non-null  int64         
 2   OrderDate          32063 non-null  datetime64[ns]
 3   StatusName         32063 non-null  object        
 4   StatusId           32063 non-null  int64         
 5   PaymentOptionName  32063 non-null  object        
 6   PaymentType        32063 non-null  int64         
 7   DespatchedCourier  32063 non-null  object        
 8   ProductsExVAT      32063 non-null  float64       
 9   DeliveryExVAT      32063 non-null  float64       
 10  VAT                32063 non-null  float64       
 11  SubTotal           32063 non-null  float64       
 12  GrandTotal         32063 non-null  float64       
 13  DateDespatched     31006 non-null  datetime64[ns]
 14  Delive

In [40]:
# Count the total number of missing values in the DataFrame
null_values_count = orders.isnull().sum().sum()

# Get columns with missing values
columns_with_null = orders.columns[orders.isnull().any()].tolist()

# Format the output message
output_message = f"The total number of null values in the DataFrame is: {null_values_count}\n"
if columns_with_null:
    output_message += f"Missing values found in the following columns: {', '.join(columns_with_null)}"
else:
    output_message += "No columns have null values."

print(output_message)

The total number of null values in the DataFrame is: 1989
Missing values found in the following columns: DateDespatched, DeliveryCost


In [41]:
# Look for null values.
orders.isnull().sum()

OrderID                 0
WebsiteID               0
OrderDate               0
StatusName              0
StatusId                0
PaymentOptionName       0
PaymentType             0
DespatchedCourier       0
ProductsExVAT           0
DeliveryExVAT           0
VAT                     0
SubTotal                0
GrandTotal              0
DateDespatched       1057
DeliveryCost          932
dtype: int64

> There are **1057 null values** in column 'DateDespatched'
>
> There are **932 null values** in column 'DeliveryCost'

In [43]:
# List of date columns you want to convert
date_columns = ['OrderDate', 'DateDespatched']

# Apply pd.to_datetime to each column in the list
for col in date_columns:
    orders[col] = pd.to_datetime(orders[col], format='%d/%m/%Y %H:%M:%S', errors='coerce')
    
# View Info
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32063 entries, 0 to 32062
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   OrderID            32063 non-null  int64         
 1   WebsiteID          32063 non-null  int64         
 2   OrderDate          32063 non-null  datetime64[ns]
 3   StatusName         32063 non-null  object        
 4   StatusId           32063 non-null  int64         
 5   PaymentOptionName  32063 non-null  object        
 6   PaymentType        32063 non-null  int64         
 7   DespatchedCourier  32063 non-null  object        
 8   ProductsExVAT      32063 non-null  float64       
 9   DeliveryExVAT      32063 non-null  float64       
 10  VAT                32063 non-null  float64       
 11  SubTotal           32063 non-null  float64       
 12  GrandTotal         32063 non-null  float64       
 13  DateDespatched     31006 non-null  datetime64[ns]
 14  Delive

In [44]:
# Normalize the datetime column to keep only the date part (set time to 00:00:00)
orders['OrderDate'] = pd.to_datetime(orders['OrderDate']).dt.normalize()

In [45]:
# Return the status of orders.
orders['StatusName'].unique().tolist()

['Dispatched', 'Deleted', 'Processing']

In [46]:
# Return the number of rows with status 'Deleted'.
print("There are", orders[orders.StatusName == 'Deleted'].shape[0],"rows with status 'Deleted'")

There are 925 rows with status 'Deleted'


In [47]:
# Delete orders with status 'Deleted'.
orders = orders[orders['StatusName'] != 'Deleted']

# Reset the index
orders.reset_index(drop=True, inplace=True)

In [48]:
# Look for null values.
orders.isnull().sum()

OrderID                0
WebsiteID              0
OrderDate              0
StatusName             0
StatusId               0
PaymentOptionName      0
PaymentType            0
DespatchedCourier      0
ProductsExVAT          0
DeliveryExVAT          0
VAT                    0
SubTotal               0
GrandTotal             0
DateDespatched       134
DeliveryCost         132
dtype: int64

> There are now **134 null values** in column 'DateDespatched'.
>
> There are **132 null values** in column 'DeliveryCost'.
> 
>`Please note that:`**These values will be removed from the dataset.**

In [50]:
# View metadata
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31138 entries, 0 to 31137
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   OrderID            31138 non-null  int64         
 1   WebsiteID          31138 non-null  int64         
 2   OrderDate          31138 non-null  datetime64[ns]
 3   StatusName         31138 non-null  object        
 4   StatusId           31138 non-null  int64         
 5   PaymentOptionName  31138 non-null  object        
 6   PaymentType        31138 non-null  int64         
 7   DespatchedCourier  31138 non-null  object        
 8   ProductsExVAT      31138 non-null  float64       
 9   DeliveryExVAT      31138 non-null  float64       
 10  VAT                31138 non-null  float64       
 11  SubTotal           31138 non-null  float64       
 12  GrandTotal         31138 non-null  float64       
 13  DateDespatched     31004 non-null  datetime64[ns]
 14  Delive

In [51]:
# Delete null values from column 'DateDespatched'
orders = orders.dropna(subset=['DateDespatched'])

# Reset the index
orders.reset_index(drop=True, inplace=True)

In [52]:
# Look for null values.
orders.isnull().sum()

OrderID              0
WebsiteID            0
OrderDate            0
StatusName           0
StatusId             0
PaymentOptionName    0
PaymentType          0
DespatchedCourier    0
ProductsExVAT        0
DeliveryExVAT        0
VAT                  0
SubTotal             0
GrandTotal           0
DateDespatched       0
DeliveryCost         0
dtype: int64

In [53]:
# View metadata
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31004 entries, 0 to 31003
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   OrderID            31004 non-null  int64         
 1   WebsiteID          31004 non-null  int64         
 2   OrderDate          31004 non-null  datetime64[ns]
 3   StatusName         31004 non-null  object        
 4   StatusId           31004 non-null  int64         
 5   PaymentOptionName  31004 non-null  object        
 6   PaymentType        31004 non-null  int64         
 7   DespatchedCourier  31004 non-null  object        
 8   ProductsExVAT      31004 non-null  float64       
 9   DeliveryExVAT      31004 non-null  float64       
 10  VAT                31004 non-null  float64       
 11  SubTotal           31004 non-null  float64       
 12  GrandTotal         31004 non-null  float64       
 13  DateDespatched     31004 non-null  datetime64[ns]
 14  Delive

In [54]:
# Rename columns
orders = orders.rename(columns={
    'OrderID': 'order_id',
    'WebsiteID': 'website_id',
    'OrderDate': 'order_date',
    'StatusName': 'status',
    'StatusId': 'status_id',
    'PaymentOptionName': 'payment_option',
    'PaymentType': 'payment_type',
    'DespatchedCourier': 'courier',
    'ProductsExVAT': 'products_ex_vat',
    'DeliveryExVAT': 'delivery_ex_vat',
    'VAT': 'vat',
    'SubTotal': 'subtotal',
    'GrandTotal': 'total',
    'DateDespatched': 'date_despatched',
    'DeliveryCost': 'delivery_cost'
})

# See updated DataFrame.
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31004 entries, 0 to 31003
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   order_id         31004 non-null  int64         
 1   website_id       31004 non-null  int64         
 2   order_date       31004 non-null  datetime64[ns]
 3   status           31004 non-null  object        
 4   status_id        31004 non-null  int64         
 5   payment_option   31004 non-null  object        
 6   payment_type     31004 non-null  int64         
 7   courier          31004 non-null  object        
 8   products_ex_vat  31004 non-null  float64       
 9   delivery_ex_vat  31004 non-null  float64       
 10  vat              31004 non-null  float64       
 11  subtotal         31004 non-null  float64       
 12  total            31004 non-null  float64       
 13  date_despatched  31004 non-null  datetime64[ns]
 14  delivery_cost    31004 non-null  float

>>**Recalculate column 'subtotal':**

In [56]:
# Calculate 'subtotal'
orders['total_price'] = orders ['products_ex_vat'] + orders['delivery_ex_vat']

# See updated DataFrame.
orders.head()

order_id  website_id order_date      status  status_id   payment_option  \
0   2729772      971172 2024-06-03  Dispatched          6           Paypal   
1   2729773      971173 2024-06-03  Dispatched          6  Payment Gateway   
2   2729774      889210 2024-06-03  Dispatched          6  Payment Gateway   
3   2729775      971174 2024-06-03  Dispatched          6  Payment Gateway   
4   2729776      971175 2024-06-03  Dispatched          6           Paypal   

   payment_type     courier  products_ex_vat  delivery_ex_vat    vat  \
0             1  DX Express            26.46             7.37   6.76   
1             2  DX Express            18.80             7.37   5.23   
2             2  DX Freight           351.07            19.99  74.22   
3             2  DX Express            56.14             0.00  11.23   
4             1  DX Freight           144.16             0.00  28.84   

   subtotal   total date_despatched  delivery_cost  total_price  
0     26.46   40.59      2024-06-03           6.95        33.83  
1     18.80   31.40      2024-06-03           6.95        26.17  
2    351.07  445.28      2024-06-03          25.33       371.06  
3     56.14   67.37      2024-06-03           6.95        56.14  
4    144.16  173.00      2024-06-03          16.87       144.16

In [57]:
# See updated DataFrame.
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31004 entries, 0 to 31003
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   order_id         31004 non-null  int64         
 1   website_id       31004 non-null  int64         
 2   order_date       31004 non-null  datetime64[ns]
 3   status           31004 non-null  object        
 4   status_id        31004 non-null  int64         
 5   payment_option   31004 non-null  object        
 6   payment_type     31004 non-null  int64         
 7   courier          31004 non-null  object        
 8   products_ex_vat  31004 non-null  float64       
 9   delivery_ex_vat  31004 non-null  float64       
 10  vat              31004 non-null  float64       
 11  subtotal         31004 non-null  float64       
 12  total            31004 non-null  float64       
 13  date_despatched  31004 non-null  datetime64[ns]
 14  delivery_cost    31004 non-null  float

In [58]:
# Delete unnessary columns
orders = orders.drop(['subtotal'], axis= 1)

# View Dataframe
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31004 entries, 0 to 31003
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   order_id         31004 non-null  int64         
 1   website_id       31004 non-null  int64         
 2   order_date       31004 non-null  datetime64[ns]
 3   status           31004 non-null  object        
 4   status_id        31004 non-null  int64         
 5   payment_option   31004 non-null  object        
 6   payment_type     31004 non-null  int64         
 7   courier          31004 non-null  object        
 8   products_ex_vat  31004 non-null  float64       
 9   delivery_ex_vat  31004 non-null  float64       
 10  vat              31004 non-null  float64       
 11  total            31004 non-null  float64       
 12  date_despatched  31004 non-null  datetime64[ns]
 13  delivery_cost    31004 non-null  float64       
 14  total_price      31004 non-null  float

In [59]:
# Save DataFrame as csv.
orders.to_csv('orders_clean.csv')

#

[Logo](https://www.kitepackaging.co.uk/)
<p style="background-color:#808080;"><span style="color:#ffffff;"><i> 
<img src="https://www.kitepackaging.co.uk/images/kite-logo-2021.png" alt="Logo" width="100" height="200">

### **3 - Customers**

In [37]:
# View the first few rows of the dataset.
customers.head()

WebsiteID DeliveryPostCode DeliveryCountry   Latitude  Longitude  \
0        245            L39 2  United Kingdom  53.561637  -2.896693   
1        597            PR3 1  United Kingdom  53.867059  -2.719333   
2        739            CV2 2  United Kingdom  52.424559  -1.463448   
3        993            W1W 8  United Kingdom  51.519216  -0.140903   
4        994           LE15 8  United Kingdom  52.656127  -0.702324   

   Distance_km  Distance_miles  
0      200.136         124.359  
1      231.076         143.584  
2       10.301           6.401  
3      149.528          92.912  
4       89.569          55.656

In [39]:
# Display summary of DataFrame.
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23178 entries, 0 to 23177
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   WebsiteID         23178 non-null  int64  
 1   DeliveryPostCode  23178 non-null  object 
 2   DeliveryCountry   23178 non-null  object 
 3   Latitude          23178 non-null  float64
 4   Longitude         23178 non-null  float64
 5   Distance_km       23178 non-null  float64
 6   Distance_miles    23178 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 1.2+ MB


In [41]:
# Count the total number of missing values in the DataFrame
null_values_count = customers.isnull().sum().sum()

# Get columns with missing values
columns_with_null = customers.columns[customers.isnull().any()].tolist()

# Format the output message
output_message = f"The total number of null values in the DataFrame is: {null_values_count}\n"
if columns_with_null:
    output_message += f"Missing values found in the following columns: {', '.join(columns_with_null)}"
else:
    output_message += "No columns have null values."

print(output_message)

The total number of null values in the DataFrame is: 0
No columns have null values.


In [43]:
# Look for null values.
customers.isnull().sum()

WebsiteID           0
DeliveryPostCode    0
DeliveryCountry     0
Latitude            0
Longitude           0
Distance_km         0
Distance_miles      0
dtype: int64

In [45]:
# Convert all strings in the 'DeliveryPostCode' column to uppercase
customers['DeliveryPostCode'] = customers['DeliveryPostCode'].str.upper()

In [47]:
# Rename selected columns.
customers = customers.rename(columns={
    'WebsiteID': 'website_id',
    'DeliveryPostCode': 'delivery_postcode',
    'DeliveryCountry': 'delivery_country',
    'Latitude':'latitude',
    'Longitude' :'longitude', 
    'Distance_km': 'distance_km',
    'Distance_miles': 'distance_miles'
})

# See updated DataFrame.
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23178 entries, 0 to 23177
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   website_id         23178 non-null  int64  
 1   delivery_postcode  23178 non-null  object 
 2   delivery_country   23178 non-null  object 
 3   latitude           23178 non-null  float64
 4   longitude          23178 non-null  float64
 5   distance_km        23178 non-null  float64
 6   distance_miles     23178 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 1.2+ MB


In [49]:
# Delele unnessary columns
customers = customers.drop('distance_km', axis= 1)

# View Dataframe
customers.head()

website_id delivery_postcode delivery_country   latitude  longitude  \
0         245             L39 2   United Kingdom  53.561637  -2.896693   
1         597             PR3 1   United Kingdom  53.867059  -2.719333   
2         739             CV2 2   United Kingdom  52.424559  -1.463448   
3         993             W1W 8   United Kingdom  51.519216  -0.140903   
4         994            LE15 8   United Kingdom  52.656127  -0.702324   

   distance_miles  
0         124.359  
1         143.584  
2           6.401  
3          92.912  
4          55.656

In [51]:
# Save DataFrame as csv.
customers.to_csv('customers_clean.csv')

#

[Logo](https://www.kitepackaging.co.uk/)
<p style="background-color:#808080;"><span style="color:#ffffff;"><i> 
<img src="https://www.kitepackaging.co.uk/images/kite-logo-2021.png" alt="Logo" width="100" height="200">

####